### DISCUSSION TITLE
NextJs build SSR instead of Static page when tRPC `ssr` is true

### DISCUSSION THREAD

**Discussion by codart1** [trpc/trpc#958] [Q&A]
I tried to implement `tRPC` on my NextJs project. Everything works well except when I try to run `next build`. It emits:
![image](https://user-images.githubusercontent.com/1023519/133823856-6aeb4b01-b0af-4b01-a798-0255a73b22f1.png)
As you can see, I expect pages like `/404`, `/register`, '/test'... to be static but they actually SSR.
I tried to turn off `ssr` option in: https://trpc.io/docs/nextjs#ssr-boolean-default-false and try to build again, this time NextJs did build static pages as I expected:
![image](https://user-images.githubusercontent.com/1023519/133824410-a80b2f31-7b43-41c9-8b3c-fd54bceb3c0b.png)

Is that a bug or by design?

**Comment by KATT** [comment]
This is by design. If you use `ssr: true` you can't have any pages being "actually" static as the magic relies on [`getInitialProps`](https://nextjs.org/docs/api-reference/data-fetching/getInitialProps).

> `getInitialProps` will disable Automatic Static Optimization.

Might be worth mentioning this in the docs! (Feel free to add a PR!)


Worth noting though is that even if you use `getInitialProps` [you can statically cache your page in Vercel](https://twitter.com/alexdotjs/status/1386274093041950722) - in tRPC you can read how to achieve that [in this section](https://trpc.io/docs/caching#app-caching). There's also a reference [example here](https://github.com/KATT/typescript.careers/blob/3ad27919af3a077d1f52551ad8dfe2b3343d166e/src/pages/_app.tsx#L95-L106), which is the source code of [typescript.careers](https://typescript.careers).

If you still prefer "real" SSG - you have to omit `ssr: true` and prefetch data manually - see docs [here](https://trpc.io/docs/ssg).

**Reply by mmkal** [reply]
This discussion might be of interest. Seems like a long shot but maybe one day SSR + SSG will be possible: https://github.com/vercel/next.js/discussions/11424
